### PR TITLE
Cog 329 incremental users index update based on cws

### PR DESCRIPTION
**Pull request by sorych**

### CODE REVIEW

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/CapUsersJobModule.java [hunk 1]
```diff
@@ -3,82 +3,109 @@
...
-      throw new UnsupportedOperationException("the job mode other that INITIAL is not supported");
+    switch (defineJobMode()) {
+      case INITIAL_LOAD:
+        configureInitialMode();
+        break;
+      case INCREMENTAL_LOAD:
+        configureIncrementalMode();
+        break;
+      default:
+        String erroeMsg = "Job mode cannot be defined";
```
Comment (sorych) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/CapUsersJobModule.java]: typo

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/CapUsersJobModule.java [hunk 2]
```diff
@@ -3,82 +3,109 @@
...
-import static gov.ca.cwds.jobs.common.mode.DefaultJobMode.INITIAL_LOAD;
+import java.time.LocalDateTime;
 
 public class CapUsersJobModule extends AbstractBaseJobModule {
   private static final Logger LOG = LoggerFactory.getLogger(CapUsersJobModule.class);
 
   private Class<? extends BulkWriter<ChangedUserDto>> capElasticWriterClass;
 
-  private Class<? extends CapUsersIterator> capUsersJobBatchIterator;
+  private Class<? extends IdmService> perryService;
```
Comment (sorych) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/CapUsersJobModule.java]: should be renamed

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersInitialJob.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import java.time.LocalDateTime;
+
+public class CapUsersInitialJob extends AbstractCapUsersJob {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CapUsersInitialJob.class);
+
+  @Override
+  void runJob() {
+    LocalDateTime dateTimeAtStart = LocalDateTime.now();
+    LOGGER.info("start Initial Cap Users Job, creating timestampSavePoint at {}", dateTimeAtStart);
+    createSavePoint(dateTimeAtStart);
```
Comment (sorych) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersInitialJob.java]: probably should be created after the batch is over.

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/CwsOffice.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.io.Serializable;
+
+
+@Entity
+@Table(name = "CWS_OFFT")
+public class CwsOffice extends CmsPersistentObject {
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/CwsOffice.java]: Are you sure this is absent in core-api libraries?

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/StaffPerson.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.io.Serializable;
+
+@Entity
+@Table(
+        name = "STFPERST"
+)
+public class StaffPerson extends CmsPersistentObject {
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/StaffPerson.java]: Probably it's better way to re-use core-api classes

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/UserId.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+
+@NamedQuery(
+        name = UserId.CWSCMS_INCREMENTAL_LOAD_QUERY_NAME,
+        query = UserId.CWS_CMS_INCREMENTAL_LOAD_QUERY
+)
+@Entity
+@Table(name = "USERID_T")
+public class UserId extends CmsPersistentObject {
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/UserId.java]: Consider using core-api

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/iterator/CapUsersInitialJobBatchIterator.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CapUsersInitialJobBatchIterator implements CapUsersIterator {
+
+  private String paginationToken;
+  private boolean eol;
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/iterator/CapUsersInitialJobBatchIterator.java]: Name might me not self-describing

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalJob.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(CapUsersIncrementalJob.class);
+
+  @Inject
+  @CmsSessionFactory
+  private SessionFactory cmsSessionFactory;
+
+
+  @Override
+  void runJob() {
+    LocalDateTime dateTimeAtStart = LocalDateTime.now();
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalJob.java]: This looks critical bug introduced. In my opinion you should save timestamp of concrete record in database to guarantee you would not lose any changes, LocalDateTime.now() Application time is the same as database server time?

**Comment by AlexanderSerbin** [comment]
I would like like to discuss using flags as well. Probably we may come up with better solution.

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/CwsOffice.java (hunk 4 above)
Comment (sorych) [reply jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/CwsOffice.java]: Yes it is absent. It is present in Perry but not all columns I need are mapped so I decided to create an own one one.

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/StaffPerson.java (hunk 5 above)
Comment (sorych) [reply jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/StaffPerson.java]: The same as Office

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/UserId.java (hunk 6 above)
Comment (sorych) [reply jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/entity/UserId.java]: the same situation as Office and StaffPerson

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalJob.java (hunk 8 above)
Comment (sorych) [reply jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalJob.java]: good point. Let's discuss

File: jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalUpdatesProcessor.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  @Inject
+  private ElasticSearchBulkCollector<ChangedUserDto> elasticSearchBulkCollector;
+
+  @Inject
+  private CwsChangedUsersService cwsChangedUsersService;
+
+  @Inject
+  private CapChangedUsersService capChangedUsersService;
+
+  public void processUpdates() {
```
Comment (AlexanderSerbin) [inline jobs-cap-users/src/main/java/gov/ca/cwds/jobs/cap/users/job/CapUsersIncrementalUpdatesProcessor.java]: Looks good

**Pull request merged**